charms/istio-pilot/src/charm.py

```diff
@@ -927,47 +894,64 @@ def _use_https(self) -> bool:
         return self._use_https_with_tls_provider() or self._use_https_with_tls_secret()
 
     def _use_https_with_tls_secret(self) -> bool:
-        """Return True if both SSL key and crt are set with save-tls-secret, False otherwise.
+        """Return True if tls-secret-id has been configured, False otherwise.
 
         Raises:
-            ErrorWithStatus: if one of the values is missing.
+            ErrorWithStatus("tls-secret-id was provided, but the secret could not be found - ...):
+                if a secret ID is passed through the tls-secret-id configuration option, but the
+                secret cannot be found.
+            ErrorWithStatus("Access to the istio-tls-secret must be granted."):
+                if the secret ID is passed, but the access to the secret is not granted.
+            ErrorWithStatus(Missing TLS value(s), please add them to the secret")
+                if any of the TLS values are missing from the secret.
         """
 
-        # Ensure the secret that holds the values exist otherwise fail as this
-        # is an error in our code
-        try:
-            secret = self.model.get_secret(label=TLS_SECRET_LABEL)
-        except SecretNotFoundError:
+        # Check if the tls-secret-id holds any value and get the secret
+        if not self._tls_secret_id:
             return False
 
-        ssl_key = secret.get_content(refresh=True)["ssl-key"]
-        ssl_crt = secret.get_content(refresh=True)["ssl-crt"]
-
-        # This block of code is more a validation than a behaviour of the charm
-        # Ideally this will never be executed, but we need a mechanism to know that
-        # these values were correctly saved in the secret
-        if _xor(ssl_key, ssl_crt):
-            missing = "ssl-key"
-            if not secret.get_content(refresh=True)["ssl-crt"]:
-                missing = "ssl-crt"
-            self.log.error(f"Missing {missing}, this is most likely an error with the charm.")
-            raise GenericCharmRuntimeError(f"Missing {missing}, cannot configure TLS.")
-        elif not ssl_key and not ssl_crt:
+        try:
+            secret = self.model.get_secret(id=self._tls_secret_id)
+        except SecretNotFoundError as secret_error:
+            if "not granted access to" in secret_error.args[0]:
```

Comment:
I have filed https://github.com/canonical/operator/issues/1229 for this because the behaviour between Harness and the actual deployed charm differs. I will change the charm code and test according to the actual behaviour, but I will skip the unit test case, otherwise it will fail.

Reply:
Sure, let's watch this so if they fix it we can update it here as well.